### ISSUE TITLE
umap.plot support for cuml.umap objects

### ISSUE THREAD

**Issue by lc5415** [lmcinnes/umap#476]
Hello! First of all thanks for this amazing package! I wanted to raise an issue that I know does not strictly involve this code base but that you may be able to help with:

I want to use the `umap.plot` functions with `cuml.UMAP` objects(I've already raised an issue on the **cuml** side. I know this may not concern but I thought you could perhaps help. As of now my code does not run if I pass a `cuml.UMAP` object the same way it would if I pass it a `umap.UMAP` object. If you have access to a GPU this code should not run for the `cuml.UMAP` object:
```python
import sklearn.datasets
import pandas as pd
import numpy as np
import umap
import cuml
import umap.plot

pendigits = sklearn.datasets.load_digits()

mapper = umap.UMAP().fit(pendigits.data)

umap.plot.points(mapper, labels = pendigits.target)

### Now cuml

mapper = cuml.UMAP().fit(pendigits.data)

umap.plot.points(mapper, labels = pendigits.target)
```

I am not able to paste the whole error log here because the GPU I have access to is from work and there is a firewall in my workplace that does not allow me to write issues or comments in GitHub. The main error I get goes something like:

```
'c' argument has 1797 elements, which is inconsistent with 'x' and 'y' with size 0
```

**Comment by lmcinnes** [comment]
I think this is actually something that can be dealt with relatively easily (I hope). As a temporary workaround you could try:

```python
mapper = cuml.UMAP().fit(pendigits.data)
mapper.embedding_ = mapper._embedding_
umap.plot.points(mapper, labels = pendigits.target)
```

If that works for you then the fix should be relatively easy and I can try to get it committed at some point soon.

**Comment by lc5415** [comment]
I got excited quickly, unfortunately, `cuml.UMAP` does not have a `._embedding_` attribute, it's got the usual `.embedding` so that, unfortunately, did not solve it. I had quick glimpse at the code for `umap.plot` and I see that it goes into the thousand(s) lines, do you think it is reasonable for someone new to this repo to be fixed?

**Comment by lmcinnes** [comment]
Sorry, I was trying to browse through their source code; perhaps they have some level of indirection? Ideally you just want an attribute ``embedding_`` that contains the actual x,y locations of the points. After that the plot package *should* just work. It doesn't really use too much magic of the UMAP objects themselves.

**Comment by lc5415** [comment]
Found the fix, it was very silly, the problem was that by default when working with `cuml` the array type is `cuml.common.array.CumlArray` and as of now the `umap.plot` function cannot handle the conversion of that to a numpy array which simply enough can be achieved like:
```python
mapper = cuml.UMAP().fit(pendigits.data)
mapper.embedding_ = mapper.embedding_.to_output('numpy')
mapper2.metric = 'euclidean' # this is to avoid a warning, as of today, cuml does not support other metrics
umap.plot.points(mapper, labels = pendigits.target)
```

Perhaps you can add a simple type check within `umap.plot` that detects whether the embedding is  a `CumlArray' and if it turns it into a `np.array`. The current supported array format by the `cuml` library are _cudf, cupy and numpy_

Reference to the metrics not being supported [source](https://docs.rapids.ai/api/cuml/stable/api.html?highlight=umap#cuml.UMAP)
> This module is heavily based on Leland McInnes’ reference UMAP package. However, there are a number of differences and features that are not yet implemented in cuml.umap:
Using a non-Euclidean distance metric (support for a fixed set of non-Euclidean metrics is planned for an upcoming release).

**Comment by lmcinnes** [comment]
Thanks -- that looks like a relatively easy fix. I will try to add a check and do the conversion if it is a CumlArray soon.

**Comment by lmcinnes** [comment]
So I started to refactor things a little. I'll add the CumlArray conversions once I figure out how to properly test for that.
